src/fields.js

```diff
@@ -0,0 +1,223 @@
+      const { children } = validate ? validate(currentValue, context) : internalState.get()
+      children.forEach(x => x.validate(addParent(context, currentValue)))
+    },
+    setSubmitted(isSubmitted) {
+      const { children } = internalState.update(x => updateState(x, { isSubmitted }))
+      children.forEach(x => x.setSubmitted(isSubmitted))
+    },
+    reset() {
+      const { children } = internalState.update(x => initialState)
+      children.forEach(x => x.reset())
+    },
+    value,
+    state: { get: internalState.get, subscribe: internalState.subscribe },
+    helpers: {
+      add(initialValue) {
+        internalState.update(x => {
+          const newChildren = [...x.children, createFormFieldsAt(initialValue, x.children.length)]
+          return updateState(x, { children: newChildren })
+        })
+      },
+      remove(entry) {
+        internalState.update(x => {
+          const newChildren = x.children.filter(x => x !== entry)
+          return updateState(x, { children: newChildren })
+        })
+      }
+    }
+  }
+
+  function createFormFieldsAt(initialValue, index) {
+    const fullName = `${name}[${index}]`
+    return createObjectFormField({
+      name: fullName,
+      initialValue,
+      field: normalize({ type: 'object', fields: field.fields }, fullName),
+    })
+  }
+}
+
+function createBasicFormField({ name, initialValue, field }) {
+
+  const initialFieldState = deriveFieldState({ value: initialValue })
+  const internalState = createState(initialFieldState)
+  const validate = bindValidate(field.validate, internalState)
+
+  const value = {
+    get() { return internalState.get().value },
+    subscribe(f) {
+      return internalState.subscribe(({ value: newValue }, { value: oldValue }) => {
+        if (newValue === oldValue) return
+        f(newValue, oldValue)
+      })
+    },
+  }
+
+  return {
+    type: 'basic',
+    name,
+    validate(context) {
+      if (validate) validate(value.get(), context)
+    },
+    setSubmitted(isSubmitted) {
+      internalState.update(x => updateState(x, { isSubmitted }))
+    },
+    reset() {
+      internalState.update(x => initialFieldState)
+    },
+    value,
+    state: { get: internalState.get, subscribe: internalState.subscribe },
+    eventHandlers: {
+      onBlur() {
+        internalState.update(x => updateState(x, { hasFocus: false, isVisited: true }))
+      },
+      onFocus() {
+        internalState.update(x => updateState(x, { hasFocus: true }))
+      },
+      onChange(eOrValue) {
+        const value = eOrValue && eOrValue.target
+          ? eOrValue.target.value
+          : eOrValue
+
+        internalState.update(x => updateState(x, { value }))
+      },
+    },
+  }
+}
+
+function updateState(fieldState, update) {
+  return deriveFieldState({ ...fieldState, ...update })
+}
+
+function deriveFieldState({
+  error = false,
+  isSubmitted = false,
+  isVisited = false,
+  hasFocus = false,
+  ...rest
+}) {
+  return {
```

Comment:
Is het handig om hier net als showError een field op te nemen om de logica in `<LabelAndError />` in te stoppen?

Reply:
Als we hier een standaard voor kunnen bedenken die overal geldt kunnen we dat wel doen ja. Ik heb nu maar gewoon een beetje gevogeld. Ik zal er nog es naar kijken met dat idee in mijn hoofd

Reply:
Ik denk dat we dit na een paar projecten nog es moeten bekijken